src/alloc/globalalloc.c

```diff
@@ -5,33 +5,145 @@ static struct allocator global_allocator;
 
 #define POOL_SIZE 16 * 1024 * 1024
 
+/** Pool node for metadata */
+struct Node
+{
+  char *pool_element;
+  int allocated; // Should be -1 if node is not a head.
+  int index;
+  bool used;
+  bool head;
+};
+
+/** Pool of metadata that maps to the pool array */
+static struct Node metadata[POOL_SIZE];
+
 /** The pool of memory that can be accessed by the global allocator */
 static char pool[POOL_SIZE];
 
 /** Initializes the global allocator */
-static void global_init() {
-    
+static void global_init()
+{
+  for (int i = 0; i < POOL_SIZE; i++)
+  {
+    metadata[i].pool_element = &pool[i];
+    metadata[i].used = false;
+    metadata[i].head = false;
+    metadata[i].allocated = -1;
+    metadata[i].index = i;
+  }
+}
+
+/** Finds corresponding metadata to a given pool element */
+static struct Node *find_element(void *p)
+{
+  struct Node *element;
```

Comment:
You should initialize `element` to be `NULL` in this case. Since you are declaring a variable on the stack (in this case, declaring a pointer on the stack), the variable could hold garbage data (since the data that is stored on the stack before use is undefined and random). By initializing as `NULL`, you remove this randomness and make sure that this function returns `NULL` whenever the function is called. In short, any variable initialized on the stack without a given value has a random variable, not 0 like in languages such as Java.